### DISCUSSION TITLE
How to get the positions and indices of a list of meshes?

### DISCUSSION THREAD

**Discussion by krispya** [donmccurdy/glTF-Transform#1170] [Q&A] [answered]
Hello! I am attempting to process a GLTF to get two buffers, one that has the positions of all the meshes joined and another that is the indices for this position buffer. The intent is to feed this into recast to get a navmesh generated on a Node server. Could you help point me in the right direction? So far I have discovered that I can get an array of all the meshes, then their primitives, then search the primitives for a position attribute.

**Comment by donmccurdy** [comment, marked as answer]
A good starting point might be:

1. remove all materials
2. remove all non-essential vertex attributes
3. flatten the node hierarchy
4. join meshes
5. retrieve vertex positions and indices

Untested code, but it should be something like:

```js
import { prune, flatten, join } from '@gltf-transform/functions';

// 1. remove all materials
for (const material of document.getRoot().listMaterials()) {
  material.dispose();
}

// 2. remove all non-essential vertex attributes
for (const mesh of document.getRoot().listMeshes()) {
  for (const prim of mesh.listPrimitives()) {
    for (const semantic of prim.listSemantics()) {
      if (semantic !== 'POSITION') {
        prim.setAttribute(semantic, null);
      }
    }
  }
}


await document.transform(
  prune(),

  // 3. flatten the node hierarchy
  flatten(),

  // 4. join meshes
  join()
);



// 5. retrieve vertex positions and indices
const prims = document.getRoot().listMeshes()
  .flatMap((mesh) => mesh.listPrimitives());

if (prims.length > 1) {
  throw new Error(`Still have ${prims.length} mesh primitives!`);
}

const position = prim.getAttribute('POSITION').getArray();
const indices = prim.getIndices().getArray();
```

There are a few additional things you may need to consider. If the mesh data is quantized and your navmesh generator can't handle that, consider using the dequantize() transform. If the remaining node in the scene has a translation/rotation/scale (I don't think it should...) you may also need to either apply that to the geometry, or use it to offset the resulting navmesh.

**Reply by krispya** [reply]
Thanks Don! I'm going to try this out and report back.

**Reply by krispya** [reply]
This worked almost as is with dequantize. However, it is very small. Maybe there is a transform on the scene that needs to be applied? Or maybe each mesh needs to have its transforms applied before being joined?

I understand how to get the matrix of the node with `root.listNodes()[1].getWorldMatrix()`, but how would I apply it to the geometry? If this feels like its own question I can start a new discussion too.

**Reply by krispya** [reply]
I ended up answering my own question. I had to loop through each vertex and apply the world matrix with `gl-matrix`.
```js
const matrix = root.listNodes()[0].getWorldMatrix();

const transformedPositions = [];

for (let i = 0; i < positions.length; i += 3) {
	let position = vec3.fromValues(positions[i], positions[i + 1], positions[i + 2]);
	let transformedPosition = vec3.transformMat4(vec3.create(), position, matrix);
	transformedPositions.push(...transformedPosition);
}
```

**Reply by donmccurdy** [reply]
That works! Can make it a bit simpler and maybe faster if you transform the primitive beforehand:

https://gltf-transform.dev/modules/functions/functions/transformPrimitive

**Reply by krispya** [reply]
I ran into a mesh where the number of primitives did not reduce to 1 after join. What does this mean?

**Reply by donmccurdy** [reply]
See:

- https://github.com/donmccurdy/glTF-Transform/blob/main/packages/functions/src/flatten.ts
- https://github.com/donmccurdy/glTF-Transform/blob/main/packages/functions/src/join.ts

Most likely it would mean that something in the scene affects a particular mesh (e.g. with animation or morph targets) or that the primitives are incompatible (e.g. with quantized vertex attributes, or uint16 vs uint32 indices). Any incompatibility would prevent merging primitives.

**Reply by krispya** [reply]
Okay it turned out to be trees that had skinning in the glTF I should be able to destroy all skins and morphs before navmesh generation, yeah? I see I can `setSkin` on a node, but how would I remove it entirely, while leaving the mesh? Also let me know if I should start a new discussion.

**Reply by donmccurdy** [reply]
`setSkin(null)` should work fine here, or to remove the skins entirely:

```js
for (const skin of document.getRoot().listSkins()) {
  skin.dispose();
}
```

Disposing the skins will automatically detach them from any meshes.
